src/platforms/esp32/components/avm_builtins/adc_driver.c

```diff
@@ -0,0 +1,771 @@
+        ESP_LOGE(TAG, "no valid adc channel resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    term channel_resource = argv[0];
+    struct ChannelResource *chan_rsrc;
+    if (UNLIKELY(!to_channel_resource(channel_resource, &chan_rsrc, ctx))) {
+        ESP_LOGE(TAG, "resource is not a valid adc channel resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    if (chan_rsrc->calibration > ESTIMATED) {
+#if ADC_CALI_SCHEME_CURVE_FITTING_SUPPORTED
+        ESP_LOGD(TAG, "deregister curve fitting calibration scheme");
+        esp_err_t err = adc_cali_delete_scheme_curve_fitting(chan_rsrc->cali_handle);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG, "failed to release calibration profile");
+            return error_return_tuple(ctx, (adc_err_to_atom_term(ctx->global, err)));
+        }
+
+#elif ADC_CALI_SCHEME_LINE_FITTING_SUPPORTED
+        ESP_LOGD(TAG, "deregister line fitting calibration scheme");
+        esp_err_t err = adc_cali_delete_scheme_line_fitting(chan_rsrc->cali_handle);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG, "failed to release calibration profile");
+            return error_return_tuple(ctx, (adc_err_to_atom_term(ctx->global, err)));
+        }
+#endif
+    }
+    chan_rsrc->cali_handle = NULL;
+    chan_rsrc->calibration = UNCALIBRATED;
+
+    return OK_ATOM;
+}
+
+static term nif_adc_sample(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+
+    term chan_term = argv[0];
+    if (UNLIKELY(!is_adc_resource(ctx->global, chan_term))) {
+        ESP_LOGE(TAG, "Invalid channel resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct ChannelResource *chan_rsrc;
+    if (UNLIKELY(!to_channel_resource(chan_term, &chan_rsrc, ctx))) {
+        ESP_LOGE(TAG, "failed to convert adc channel resource");
+        RAISE_ERROR(BADARG_ATOM);
+        return BADARG_ATOM;
+    }
+
+    term unit_term = argv[1];
+    if (UNLIKELY(!is_adc_resource(ctx->global, unit_term))) {
+        ESP_LOGE(TAG, "handle supplied is not a valid adc resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct UnitResource *unit_rsrc = NULL;
+    if (UNLIKELY(!to_unit_resource(unit_term, &unit_rsrc, ctx))) {
+        ESP_LOGE(TAG, "resource supplied is not a valid adc unit resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    adc_oneshot_unit_handle_t unit_handle = NULL;
+    if (chan_rsrc->adc_unit == ADC_UNIT_1) {
+        unit_handle = unit_rsrc->unit_handle;
+#ifdef CONFIG_AVM_ADC2_ENABLE
+    } else if (chan_rsrc->adc_unit == ADC_UNIT_2) {
+        unit_handle = unit_rsrc->unit2_handle;
+#endif
+    } else {
+        ESP_LOGE(TAG, "no valid unit handle found in resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    term read_options = argv[2];
+    VALIDATE_VALUE(read_options, term_is_list);
+    term samples = interop_kv_get_value_default(read_options, ATOM_STR("\x7", "samples"), term_from_int32(DEFAULT_SAMPLES), ctx->global);
+    int samples_val = term_to_int32(samples);
```

Comment:
We are missing validity checks here. The default values guarantees that a value is always assigned, but it doesn't guarantee us that the value will be correct too.
Please, let's check they are integer, and let's also check the are in the accept valid range (e.g. what happens if a negative number is provided?)

Reply:
Indeed. That was an oversight, I try to be careful about always validating input parameters, I don't know why I forgot to here. I even did the calculation to determine the maximum number of samples that can be taken without overflowing the int that holds the accumulated results before diving by the number of samples. I don't know why I didn't just use that as a failsafe (except that over 62000 samples seemed an unrealistic concern).